zello: stabilize websocket, serialize auth, add talker/duration logs, and de-noise output

Mason10198 opened the pull request:
### Summary

This PR improves link stability and clarity of logs for the Zello bridge. It removes redundant pinging, prevents overlapping re-auth attempts, handles forced server disconnects cleanly, and adds concise “Keyed/UnKeyed” logs—including talker username (RX) and keyed duration (RX & USRP TX). It also reduces log noise at INFO level. **Additionally, it fixes intermittent `invalid stream id` errors by resetting TX state on websocket loss and never sending audio frames until a fresh `stream_id` is established.**

**New in this update:** adaptive server-driven backoffs (“woodpecker prohibited” and “empty message”), channel-online gating with a short post-login cool-down, an auth watchdog to clear stuck logins, de-duplicated “TX skip” debug reasons, periodic TX frame summaries, cached private key, OPUS tuning via env, and clean shutdown stats.

### What changed (high level)

* **Keepalive / Heartbeat**

  * Continue relying on `aiohttp`’s `autoping=True` with `heartbeat=30.0` instead of a custom ping task.
  * Enable `TCP_NODELAY` while preserving TCP keepalive options.

* **Auth & Token**

  * All token math uses **UTC** and `total_seconds()`.
  * A single-flight **`auth_lock`** ensures only one `authenticate()` runs at a time.
  * Track the **auth request `seq`** and clear `auth_in_progress` only when a matching `success` arrives.
  * Add an **auth watchdog** (`AUTH_WATCHDOG_SEC`) in monitor and TX paths to recover from stuck auth.
  * When a `refresh_token` is received, parse its **`exp`** to update `_token_expiry` and avoid premature re-auth cycles.
  * Centralized **`_reset_connection_state()`** used on closure/error to avoid stale flags.

* **Channel readiness & TX gating**

  * Gate TX on **channel status** (`on_channel_status` → `status == "online"`).
  * Add a small **post-login cool-down** (`POST_LOGIN_COOLDOWN_SEC`) before allowing TX.
  * If server returns **`channel is not ready`**, apply a short **backoff** (`CHANNEL_NOT_READY_BACKOFF_SEC`) and defer `start_stream`.

* **Adaptive backoffs (server feedback)**

  * **`woodpecker prohibited`** → exponential backoff (up to 8s) and immediate local unkey if active TX.
  * **`empty message`** → short, doubling backoff (up to 8s) to avoid rapid retries.

* **Stream lifecycle robustness**

  * Clear `_stream_id` on stop and before a new start to avoid reuse.
  * `start_tx()` waits (up to \~2s) for **`stream_id`**; if missing, sets a short retry window instead of sending frames.
  * Never send audio unless `_stream_id` is valid; include **packet id** (`_pkt_id`) in frames.
  * If the websocket drops while TX is active, immediately unkey locally (log duration), clear TX state, and **do not** attempt `stop_stream` on a dead socket.

* **Logging & observability**

  * Only log **“Logged in!”** on state change `False → True`.
  * Promote concise operational lines at **INFO**:

    * RX: `Keyed:<username>` / `UnKeyed:<username> (<secs>s)`
    * TX (USRP): `Keyed:USRP` / `UnKeyed:USRP (<secs>s)`
  * Demote noisy details to **DEBUG**:

    * Raw `start_stream`/`stop_stream` JSON (with **sensitive fields redacted**: `password`, `auth_token`, `refresh_token`)
    * Per-packet logs and frame metrics
  * De-duplicated **“TX skip: …”** reasons at DEBUG to avoid log floods.
  * Periodic **TX frame summary** (pps, avg size, total bytes) emitted roughly once per second while TXing.

* **Performance & housekeeping**

  * Micro-yields for smooth scheduling: `MAIN_LOOP_YIELD_SEC`, with small idle sleep while PTT is clear (`PTT_IDLE_SLEEP_SEC`).
  * Precompute and reuse the **OPUS codec header**; **cache the private key** to avoid repeated disk reads.
  * Optional OPUS tuning via env vars (best-effort): `OPUS_COMPLEXITY`, `OPUS_BITRATE`.
  * Graceful **`shutdown()`** cancels tasks, closes the socket, and prints final **stats** (`start_attempts`, `start_ok`, `chn_not_ready`, `read_timeouts`).

* **TX resilience on reconnect (fixes `invalid stream id`)**

  * If the websocket drops while TX is active, immediately unkey locally (log duration), clear `_txing`, `_usrp_tx_start`, and `_stream_id`, without attempting to send a `stop_stream` on a dead socket.
  * Guard before sending frames: **do not send** unless `_stream_id` is present (i.e., after a fresh `start_stream` succeeds).

### Before vs After (concise; sensitive fields redacted)

**Before**

```
Sep 06 13:54:50 systemd[1]: Started asl-zello-bridge.service.
Sep 06 13:54:52 __main__:Initialising Zello
Sep 06 13:54:52 __main__:Initialising USRP
Sep 06 13:54:52 USRPController:USRP RX gain: 0dB = 1.0
Sep 06 13:54:52 USRPController:USRP TX gain: 0dB = 1.0
Sep 06 13:54:52 ZelloController:Connecting to wss://zello.io/ws
Sep 06 13:54:52 ZelloController:Monitor task starting
Sep 06 13:54:52 ZelloController:Ping task starting
Sep 06 13:54:52 ZelloController:Authenticating with new token
Sep 06 13:54:52 ZelloController:Private key detected, getting Zello Free token
Sep 06 13:54:53 ZelloController:{"command":"logon","seq":0,"username":"[redacted]","password":"[redacted]","auth_token":"[redacted]"}
Sep 06 13:54:53 ZelloController:WSMessage(... data='{"refresh_token":"[redacted]","success":true,"seq":0}')
Sep 06 13:54:53 ZelloController:Authentication successful!
Sep 06 13:54:53 ZelloController:Logged in!
Sep 06 13:54:53 ZelloController:WSMessage(... data='{"command":"on_channel_status",...}')
Sep 06 13:54:53 ZelloController:Logged in!
Sep 06 13:54:53 ZelloController:WS closed!
... (raw start/stop JSON at INFO)
Sep 06 13:59:12 ZelloController:WSMessage(... data='{"command":"on_stream_start","from":"[redacted]"}')
Sep 06 13:59:12 ZelloController:on_stream_start
... (per-packet size at INFO)
Sep 06 13:59:13 ZelloController:WSMessage(... data='{"command":"on_stream_stop","stream_id":...}')
Sep 06 13:59:13 ZelloController:on_stream_stop
Sep 06 13:59:13 ZelloController:Logged in!
```

**After**

```
Sep 06 13:59:56 ZelloController:Connecting to wss://zello.io/ws
Sep 06 13:59:56 ZelloController:Monitor task starting
Sep 06 13:59:56 ZelloController:Authenticating with new token
Sep 06 13:59:56 ZelloController:Private key detected, getting Zello Free token
Sep 06 13:59:56 ZelloController:Logging in...
Sep 06 13:59:57 ZelloController:Authentication successful!
Sep 06 13:59:57 ZelloController:Logged in!
Sep 06 14:00:38 ZelloController:Keyed:USRP
Sep 06 14:00:42 ZelloController:UnKeyed:USRP (4.6s)
Sep 06 14:00:49 ZelloController:Keyed:[redacted]
Sep 06 14:00:51 ZelloController:UnKeyed:[redacted] (1.2s)
# (no 'invalid stream id' during reconnects; server-driven backoffs honored; TX skips logged at DEBUG)
```

### Why this is better

* **Less noise, more signal:** Clean, human-friendly `Keyed/UnKeyed` lines with durations you can scan quickly during incidents.
* **Fewer disconnect blips:** Letting `aiohttp` manage heartbeat avoids double-pinging and stray closures; adaptive backoffs respect server guidance.
* **No overlapping auth:** Single-flight auth + watchdog prevents race conditions and long-lived half-auth states.
* **Safer stream lifecycle:** Clear `stream_id` avoids accidental reuse; frames are never sent without a valid `stream_id`, eliminating `invalid stream id` errors.
* **Sharper debugging:** Redacted JSON at DEBUG, de-duplicated “TX skip” reasons, per-second TX summaries, and end-of-run stats.

### Implementation notes

* Websocket connect: `autoping=True, heartbeat=30.0`
* New/updated fields include backoff trackers (`_woodpecker_until`, `_empty_msg_backoff_until`, `_channel_backoff_until`, `_start_retry_after`), auth timers/seq (`_auth_in_progress`, `_auth_started_at`, `_auth_seq`), frame counters, cached key, prebuilt codec header, packet id counter.
* Start/Stop JSON remains available at DEBUG for deep diagnostics.
* Env var expectations unchanged; optional OPUS tuning via `OPUS_COMPLEXITY` and `OPUS_BITRATE`.

Mason10198 committed:
zello: stabilize websocket, serialize auth, add talker/duration logs, and de-noise output

- Use aiohttp autoping/heartbeat; remove custom ping from run()
- Serialize authentication with an auth_lock to prevent races
- Update _token_expiry from refresh_token exp (UTC, total_seconds)
- Log "Logged in!" only on state transition
- Demote raw start/stop JSON and "WS closed!" to DEBUG
- Track talker on RX: log "Keyed:<user>" and "UnKeyed:<user> (<secs>s)"
- Track local TX (USRP): log "Keyed:USRP" and duration on unkey
- Clear stale stream_id on stop and before start
- Handle server "on_error: kicked" as terminal (clean reconnect)
- TCP_NODELAY in addition to keepalive

mattmelling commented:
This looks great. As discussed in the previous PR I'd like to run this on a high traffic bridge for a couple of days and see how it performs.

Mason10198 commented:
Absolutely! I'm doing this on my end as well. After the initial PR I found one small bug with `invalid stream id` errors and updated my commit. Now a few hours later I've yet to see any errors. Some infrequent and breif audio drops/buffers between Zello and ASL, but no errors from the bridge service or from Asterisk, so I think I'm going to chock those up to (hopefully) just "expected internet stuff".

Mason10198 committed:
Merge branch 'mattmelling:master' into zello-stability-logging

Mason10198 committed:
zello: add adaptive backoffs, auth watchdog, TX summaries, and clean shutdown

- Add adaptive backoffs:
  • Handle "woodpecker prohibited" with exponential backoff (up to 8s).
  • Handle "empty message" with doubling backoff (up to 8s).
  • Handle "channel is not ready" with short retry backoff before TX.

- Auth improvements:
  • Introduce watchdog (AUTH_WATCHDOG_SEC) to clear stuck auth.
  • Track auth seq and clear _auth_in_progress only on matching success.
  • Centralized _reset_connection_state() for socket closures/errors.

- TX robustness:
  • Gate TX on channel "online" status and post-login cooldown.
  • Wait for stream_id with timeout before sending frames.
  • Add packet id counter (_pkt_id) for outgoing frames.
  • Per-second DEBUG TX frame summary (pps, avg size, bytes).

- Logging / observability:
  • Redact password/auth/refresh tokens in JSON logs.
  • De-duplicate "TX skip" reasons at DEBUG.
  • Log UnKeyed:USRP with duration on PTT release.
  • Emit clean shutdown stats (start_attempts, start_ok, chn_not_ready, read_timeouts).

- Performance / housekeeping:
  • Cache private key in memory.
  • Precompute/reuse codec header.
  • Allow OPUS_COMPLEXITY and OPUS_BITRATE via env.
  • Add micro-sleeps (PTT_IDLE_SLEEP_SEC, MAIN_LOOP_YIELD_SEC) to reduce CPU.

Overall: tighter gating around TX, smoother backoff behavior, safer auth state handling,
and more actionable DEBUG/INFO logs without excess noise.

Mason10198 committed:
Document log level option in README

Mason10198 commented:
I've been running this for a few days and have not seen any issues. I'm going to keep testing, but it feels stable enough to share.

Here are my debug logs for the last ~24hrs.

[asl-zello-bridge.9-13-2025.log](https://github.com/user-attachments/files/22311582/asl-zello-bridge.9-13-2025.log)

mattmelling commented:
Fantastic work, Mason. Have been testing here for a few days and it seems solid.
